api/users_test.go

```diff
@@ -0,0 +1,167 @@
+					var internaleErrorException cognitoidentityprovider.InternalErrorException
+					internaleErrorException.Message_ = &userException
+					internaleErrorException.RespMetadata.StatusCode = http.StatusInternalServerError
+		
+					return nil, &internaleErrorException
+				},
+				http.StatusInternalServerError,
+			},
+		}
+
+		for _, tt := range adminCreateUsersTests {
+			m.AdminCreateUserFunc = tt.createUsersFunction
+			api, _ := Setup(ctx, routeMux, m, poolId, clientId, clientSecret)
+
+			postBody := map[string]interface{}{"username": name, "email": email}
+	
+			body, _ := json.Marshal(postBody)
+	
+			r := httptest.NewRequest(http.MethodPost, usersEndPoint, bytes.NewReader(body))
+	
+			w := httptest.NewRecorder()
+	
+			api.Router.ServeHTTP(w, r)
+	
+			So(w.Code, ShouldEqual, tt.httpResponse)
+		}
+	})
+
+	Convey("Admin create user returns 500: error unmarshalling request body", t, func() {
+		r := httptest.NewRequest(http.MethodPost, usersEndPoint, bytes.NewReader(nil))
+
+		w := httptest.NewRecorder()
+
+		api, _ := Setup(ctx, routeMux, m, poolId, clientId, clientSecret)
+
+		api.Router.ServeHTTP(w, r)
+
+		errorBody, _ := ioutil.ReadAll(w.Body)
+		var e models.ErrorStructure
+		json.Unmarshal(errorBody, &e)
+
+		So(w.Code, ShouldEqual, http.StatusInternalServerError)
+		So(e.Errors[0].Message, ShouldEqual, "api endpoint POST user returned an error unmarshalling request body")
+	})
+
+	Convey("Validation fails 400: validating email and username throws validation errors", t, func() {
+		userValidationTests := []struct {
+			userDetails map[string]interface{}
+			errorMessage []string
+			httpResponse int
+		}{
+			// missing username
+			{
+				map[string]interface{}{"username": "", "email": email},
+				[]string{
+					apierrors.InvalidUserNameMessage,
+				},		
+				http.StatusBadRequest,
+			},
+			// missing email
+			{
+				map[string]interface{}{"username": name, "email": ""},
+				[]string{
+					apierrors.InvalidErrorMessage,
+				},
+				http.StatusBadRequest,
+			},
+			// missing both username and email
+			{
+				map[string]interface{}{"username": "", "email": ""},
+				[]string{
+					apierrors.InvalidUserNameMessage,
+					apierrors.InvalidErrorMessage,
+				},
+				http.StatusBadRequest,
+			},
+		}
+	
+		for _, tt := range userValidationTests {
+			body, _ := json.Marshal(tt.userDetails)
+	
+			r := httptest.NewRequest(http.MethodPost, usersEndPoint, bytes.NewReader(body))
+	
+			w := httptest.NewRecorder()
+	
+			api, _ := Setup(ctx, routeMux, m, poolId, clientId, clientSecret)
+
+			api.Router.ServeHTTP(w, r)
+	
+			errorBody, _ := ioutil.ReadAll(w.Body)
+			var e models.ErrorStructure
+			json.Unmarshal(errorBody, &e)
+	
+			So(w.Code, ShouldEqual, tt.httpResponse)
+			So(len(e.Errors), ShouldEqual, len(tt.errorMessage))
+			So(e.Errors[0].Message, ShouldEqual, tt.errorMessage[0])
+			if len(e.Errors) > 1 {
+				So(e.Errors[1].Message, ShouldEqual, tt.errorMessage[1])
+			}
+		}
+	})
+}
```

Comment:
If you run "Go Fmt Project" in the Goland tools then it should add a new line here. I think there are other files missing new lines at the end that it will also fix.

Reply:
Lovely - good idea. Will run that.